### PR TITLE
Alpha complex python enhancement

### PR DESCRIPTION
**Pull request by VincentRouvreau**
* fast and exact versions
* Delaunay complex version

### CODE REVIEW

File: src/python/gudhi/alpha_complex.pyx [hunk 1]
```diff
@@ -53,10 +53,12 @@ cdef class AlphaComplex:
...
     """
 
-    cdef Alpha_complex_interface * thisptr
+    cdef Alpha_complex_interface * this_ptr
+    cdef bool fast
+    cdef bool exact
 
     # Fake constructor that does nothing but documenting the constructor
-    def __init__(self, points=None, off_file=''):
+    def __init__(self, points=None, off_file='', complexity='safe'):
```
Comment (mglisse) [inline src/python/gudhi/alpha_complex.pyx:61]: I see why you would call it complexity, but I am not too fond of the name. Precision? I am open to other suggestions, or even to keep complexity if someone else likes it.

File: src/python/gudhi/alpha_complex.pyx [hunk 2]
```diff
@@ -66,15 +68,21 @@ cdef class AlphaComplex:
...
         :type off_file: string
+
+        :param complexity: Alpha complex complexity can be 'fast', 'safe' or 'exact'. Default is 'safe'.
+        :type complexity: string
         """
 
     # The real cython constructor
-    def __cinit__(self, points = None, off_file = ''):
+    def __cinit__(self, points = None, off_file = '', complexity = 'safe'):
+        assert complexity == 'fast' or complexity == 'safe' or complexity == 'exact', "Alpha complex complexity can be 'fast', 'safe' or 'exact'"
```
Comment (mglisse) [inline src/python/gudhi/alpha_complex.pyx:78]:
```suggestion
        assert complexity in ['fast', 'safe', 'exact'], "Alpha complex complexity can only be 'fast', 'safe' or 'exact'"
```

File: src/python/include/Alpha_complex_interface.h [hunk 3]
```diff
@@ -23,45 +23,74 @@
...
-  Alpha_complex_interface(const std::vector<std::vector<double>>& points) {
-    auto mkpt = [](std::vector<double> const& vec){
-      return Point_d(vec.size(), vec.begin(), vec.end());
-    };
-    alpha_complex_ = new Alpha_complex<Dynamic_kernel>(boost::adaptors::transform(points, mkpt));
+  template <typename CgalPointType>
+  std::vector<double> pt_cgal_to_cython(CgalPointType& ph) {
+    std::vector<double> vd;
+    for (auto coord = ph.cartesian_begin(); coord != ph.cartesian_end(); coord++)
+      vd.push_back(CGAL::to_double(*coord));
```
Comment (mglisse) [inline src/python/include/Alpha_complex_interface.h:43]: I wonder if we should call CGAL::exact(ph) at the beginning of this function, in the 'exact' case.
What is the 'h' for in 'ph'?

File: src/python/gudhi/alpha_complex.pyx [hunk 4]
```diff
@@ -53,10 +53,12 @@ cdef class AlphaComplex:
 
     """
 
-    cdef Alpha_complex_interface * thisptr
+    cdef Alpha_complex_interface * this_ptr
+    cdef bool fast
```
Comment (mglisse) [inline src/python/gudhi/alpha_complex.pyx:57]: Isn't that redundant with `this_ptr->fast_version_`?

File: src/python/doc/alpha_complex_user.rst [hunk 5]
```diff
@@ -16,8 +16,25 @@ Definition
...
+* When an :math:`\alpha`-complex is constructed with an infinite value of :math:`\alpha^2`, the complex is a Delaunay
+  complex (with special filtration values). The Delaunay complex without filtration values is also available by
+  passing :code:`default_filtration_value = True` to :func:`~gudhi.AlphaComplex.create_simplex_tree`.
+* For people only interested in the topology of the Alpha complex (for instance persistence), Alpha complex is
+  equivalent to the `Čech complex <https://gudhi.inria.fr/doc/latest/group__cech__complex.html>`_ and much smaller if
+  you do not bound the radii. `Čech complex <https://gudhi.inria.fr/doc/latest/group__cech__complex.html>`_ can still
+  make sense in higher dimension precisely because you can bound the radii.
+* Using the default :code:`complexity = 'safe'` makes the construction safe.
+  If you pass :code:`complexity = 'exact'` to :func:`~gudhi.AlphaComplex.__init__`, the filtration values are the exact
+  ones converted to the filtration value type of the simplicial complex. This can be very slow.
```
Comment (mglisse) [inline src/python/doc/alpha_complex_user.rst:28]: Here, I think that's just `float`.

File: src/python/doc/alpha_complex_user.rst [hunk 6]
```diff
@@ -16,8 +16,25 @@ Definition
...
+  If you pass :code:`complexity = 'safe'` (the default) or :code:`complexity = 'fast'`, the filtration values are only
+  guaranteed to have a small multiplicative error compared to the exact value, see
```
Comment (mglisse) [inline src/python/doc/alpha_complex_user.rst:30]: Uh? With 'fast' there is no guarantee whatsoever.

File: src/python/doc/alpha_complex_user.rst (hunk 5 above)
Comment (VincentRouvreau) [reply src/python/doc/alpha_complex_user.rst]: Fixed on 841dac59

File: src/python/doc/alpha_complex_user.rst [hunk 7]
```diff
@@ -16,8 +16,25 @@ Definition
...
+  `CGAL::Lazy_exact_nt<NT>::set_relative_precision_of_to_double <https://doc.cgal.org/latest/Number_types/classCGAL_1_1Lazy__exact__nt.html>`_
```
[the remark this answers is not in the document]
Comment (VincentRouvreau) [reply src/python/doc/alpha_complex_user.rst]: You are right, done on c8f5c9fc

File: src/python/doc/alpha_complex_user.rst (hunk 6 above)
Comment (VincentRouvreau) [reply src/python/doc/alpha_complex_user.rst]: No guarantee on fast, you are right: c8f5c9fc

File: src/python/gudhi/alpha_complex.pyx (hunk 4 above)
Comment (VincentRouvreau) [reply src/python/gudhi/alpha_complex.pyx]: Yes, no need to store this information. Modified on 58cb456a

File: src/python/include/Alpha_complex_interface.h (hunk 3 above)
Comment (VincentRouvreau) [reply src/python/include/Alpha_complex_interface.h:43]: I renamed ph as point on e7ccfabc
How could I check if I need to call CGAL::exact(point) ?

File: src/python/gudhi/alpha_complex.pyx (hunk 1 above)
Comment (VincentRouvreau) [reply src/python/gudhi/alpha_complex.pyx]: I am ok with your proposal, I renamed it precision on 4f02ba9c

File: src/python/gudhi/alpha_complex.pyx [hunk 8]
```diff
@@ -66,15 +67,21 @@ cdef class AlphaComplex:
...
+        :param precision: Alpha complex precision can be 'fast', 'safe' or 'exact'. Default is 'safe'.
+        :type precision: string
         """
 
     # The real cython constructor
-    def __cinit__(self, points = None, off_file = ''):
+    def __cinit__(self, points = None, off_file = '', precision = 'safe'):
+        assert precision in ['fast', 'safe', 'exact'], "Alpha complex precision can only be 'fast', 'safe' or 'exact'"
+        cdef bool fast = precision == 'fast'
+        self.exact = precision == 'safe'
```
Comment (mglisse) [inline src/python/gudhi/alpha_complex.pyx:79]: This line is a bit confusing. What about the case `precision == 'exact'`?

File: src/python/include/Alpha_complex_interface.h (hunk 3 above)
Comment (mglisse) [reply src/python/include/Alpha_complex_interface.h:43]: Ah, this function is only ever used for the input points, so there is no need to call CGAL::exact.

File: src/python/gudhi/alpha_complex.pyx (hunk 8 above)
Comment (VincentRouvreau) [reply src/python/gudhi/alpha_complex.pyx]: Oops... This fault was terrible...
I should think of a way to test exact version.

**Pull request merged**